assets/js/components/OptIn.js

```diff
@@ -21,119 +21,75 @@
  */
 import PropTypes from 'prop-types';
 import classnames from 'classnames';
+
 /**
  * WordPress dependencies
  */
-import apiFetch from '@wordpress/api-fetch';
-import { Component } from '@wordpress/element';
+import { useCallback } from '@wordpress/element';
 import { __, sprintf } from '@wordpress/i18n';
 
 /**
  * Internal dependencies
  */
-import { getMetaKeyForUserOption, sanitizeHTML } from '../util';
+import Data from 'googlesitekit-data';
+import { CORE_USER } from '../googlesitekit/datastore/user/constants';
+import { sanitizeHTML } from '../util';
+import { toggleTracking, trackEvent } from '../util/tracking';
 import Checkbox from './Checkbox';
-import {
-	isTrackingEnabled,
-	toggleTracking,
-	trackEvent,
-} from '../util/tracking';
-
-class OptIn extends Component {
-	constructor( props ) {
-		super( props );
-
-		this.state = {
-			optIn: isTrackingEnabled(),
-			error: false,
-		};
+const { useSelect, useDispatch } = Data;
 
-		this.handleOptIn = this.handleOptIn.bind( this );
-	}
+export default function OptIn( { id, name, className, optinAction } ) {
+	const enabled = useSelect( ( select ) => select( CORE_USER ).isTrackingEnabled() );
+	const saving = useSelect( ( select ) => select( CORE_USER ).isSavingUserTracking() );
+	const error = useSelect( ( select ) => select( CORE_USER ).getErrorForAction( 'saveUserTracking', [ ! enabled ] ) );
 
-	async handleOptIn( e ) {
+	const { saveUserTracking } = useDispatch( CORE_USER );
+	const handleOptIn = useCallback( ( e ) => {
 		const checked = !! e.target.checked;
-		const trackingUserOptInKey = getMetaKeyForUserOption( 'googlesitekit_tracking_optin' );
 
 		toggleTracking( checked );
-
 		if ( checked ) {
-			await trackEvent( 'tracking_plugin', this.props.optinAction );
-		}
-
-		try {
-			await apiFetch( {
-				path: '/wp/v2/users/me',
-				method: 'POST',
-				data: {
-					meta: {
-						[ trackingUserOptInKey ]: checked,
-					},
-				},
-			} );
-			this.setState( {
-				optIn: checked,
-				error: false,
-			} );
-		} catch ( err ) {
-			this.setState( {
-				optIn: ! checked,
-				error: {
-					errorCode: err.code,
-					errorMsg: err.message,
-				},
-			} );
+			trackEvent( 'tracking_plugin', optinAction );
 		}
-	}
-
-	render() {
-		const {
-			optIn,
-			error,
-		} = this.state;
 
-		const {
-			id,
-			name,
-			className,
-		} = this.props;
+		saveUserTracking( checked );
```

Comment:
We need to look at the response value here. If there is an error and the tracking value wasn't saved, we need to re-set `toggleTracking` to the previous checked value, since that part is decoupled from the datastore, since it should be able to run agnostically without all of React etc.

Maybe it would even be cleaner to do that as a `useEffect` based on `enabled`, i.e. handle the `toggleTracking` and `trackEvent` side effects based on value changes of that (since it will only change when the request is successful). If we do that, we just need to make sure it doesn't fire on the _initial_ call when data was just loaded.

Reply:
Updated to use the response of the saving tracking enabled response.